File: core/src/ten_runtime/binding/nodejs/interface/extension.ts
//
// Copyright © 2025 Agora
// This file is part of TEN Framework, an open source project.
// Licensed under the Apache License, Version 2.0, with certain conditions.
// Refer to the "LICENSE" file in the root directory for more information.
//
import type { Cmd } from "./msg/cmd/cmd.js";
import type { Data } from "./msg/data.js";
import type { AudioFrame } from "./msg/audio_frame.js";
import type { VideoFrame } from "./msg/video_frame.js";
import type { TenEnv } from "./ten_env.js";
import ten_addon from "./ten_addon.js";

export abstract class Extension {
  constructor(name: string) {
    ten_addon.ten_nodejs_extension_create(this, name);
  }

  private async onConfigureProxy(tenEnv: TenEnv): Promise<void> {
    await this.onConfigure(tenEnv);

    ten_addon.ten_nodejs_ten_env_on_configure_done(tenEnv);
  }

  private async onInitProxy(tenEnv: TenEnv): Promise<void> {
    await this.onInit(tenEnv);

    ten_addon.ten_nodejs_ten_env_on_init_done(tenEnv);
  }

  private async onStartProxy(tenEnv: TenEnv): Promise<void> {
    await this.onStart(tenEnv);

    ten_addon.ten_nodejs_ten_env_on_start_done(tenEnv);
  }

  private async onStopProxy(tenEnv: TenEnv): Promise<void> {
    await this.onStop(tenEnv);

    ten_addon.ten_nodejs_ten_env_on_stop_done(tenEnv);
  }

  private async onDeinitProxy(tenEnv: TenEnv): Promise<void> {
    await this.onDeinit(tenEnv);

    ten_addon.ten_nodejs_ten_env_on_deinit_done(tenEnv);
  }

  private async onCmdProxy(tenEnv: TenEnv, cmd: Cmd): Promise<void> {
    await this.onCmd(tenEnv, cmd);
  }

  private async onDataProxy(tenEnv: TenEnv, data: Data): Promise<void> {
    await this.onData(tenEnv, data);
  }

  private async onAudioFrameProxy(
    tenEnv: TenEnv,
    frame: AudioFrame,
  ): Promise<void> {
    await this.onAudioFrame(tenEnv, frame);
  }

  private async onVideoFrameProxy(
    tenEnv: TenEnv,
    frame: VideoFrame,
  ): Promise<void> {
    await this.onVideoFrame(tenEnv, frame);
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  async onConfigure(tenEnv: TenEnv): Promise<void> {
    // stub for override
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  async onInit(tenEnv: TenEnv): Promise<void> {
    // stub for override
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  async onStart(tenEnv: TenEnv): Promise<void> {
    // stub for override
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  async onStop(tenEnv: TenEnv): Promise<void> {
    // stub for override
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  async onDeinit(tenEnv: TenEnv): Promise<void> {
    // stub for override
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  async onCmd(tenEnv: TenEnv, cmd: Cmd): Promise<void> {
    // stub for override
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  async onData(tenEnv: TenEnv, data: Data): Promise<void> {
    // stub for override
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  async onAudioFrame(tenEnv: TenEnv, frame: AudioFrame): Promise<void> {
    // stub for override
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  async onVideoFrame(tenEnv: TenEnv, frame: VideoFrame): Promise<void> {
    // stub for override
  }
}

File: tests/ten_runtime/integration/python/subgraph_python/subgraph_python_app/ten_packages/extension/default_extension_python/extension.py
#
# Copyright © 2025 Agora
# This file is part of TEN Framework, an open source project.
# Licensed under the Apache License, Version 2.0, with certain conditions.
# Refer to the "LICENSE" file in the root directory for more information.
#
from ten_runtime import (
    AsyncExtension,
    AsyncTenEnv,
    Data,
    Cmd,
    LogLevel,
    CmdResult,
    StatusCode,
)


class FunctionEntryExtension(AsyncExtension):
    def __init__(self, name: str) -> None:
        super().__init__(name)
        self.received_cmd = None

    async def on_cmd(self, ten_env: AsyncTenEnv, cmd: Cmd) -> None:
        cmd_name = cmd.get_name()
        ten_env.log(
            LogLevel.INFO, f"function entry extension received cmd: {cmd_name}"
        )

        base, error = await ten_env.get_property_int("base")
        if error is not None:
            ten_env.log(LogLevel.ERROR, f"failed to get base: {error}")
            await ten_env.return_result(CmdResult.create(StatusCode.ERROR, cmd))
            return

        self.received_cmd = cmd

        data = Data.create("data")
        data.set_property_int("data", base)
        await ten_env.send_data(data)

    async def on_data(self, ten_env: AsyncTenEnv, data: Data) -> None:
        number, error = data.get_property_int("data")
        if error is not None:
            ten_env.log(LogLevel.ERROR, f"failed to get data: {error}")
            return

        if self.received_cmd is None:
            ten_env.log(LogLevel.ERROR, "received data but no cmd received")
            return

        cmd_result = CmdResult.create(StatusCode.OK, self.received_cmd)
        cmd_result.set_property_int("detail", number)
        await ten_env.return_result(cmd_result)


class PowerExtension(AsyncExtension):
    async def on_data(self, ten_env: AsyncTenEnv, data: Data) -> None:
        number, error = data.get_property_int("data")
        if error is not None:
            ten_env.log(LogLevel.ERROR, f"failed to get data: {error}")
            return

        power, error = await ten_env.get_property_int("power")
        if error is not None:
            ten_env.log(LogLevel.ERROR, f"failed to get power: {error}")
            return

        result = number**power

        data.set_property_int("data", result)
        await ten_env.send_data(data)


class MultiExtension(AsyncExtension):
    async def on_data(self, ten_env: AsyncTenEnv, data: Data) -> None:
        number, error = data.get_property_int("data")
        if error is not None:
            ten_env.log(LogLevel.ERROR, f"failed to get data: {error}")
            return

        multi, error = await ten_env.get_property_int("multi")
        if error is not None:
            ten_env.log(LogLevel.ERROR, f"failed to get multi: {error}")
            return

        result = number * multi
        data.set_property_int("data", result)
        await ten_env.send_data(data)


class SubstractExtension(AsyncExtension):
    async def on_data(self, ten_env: AsyncTenEnv, data: Data) -> None:
        number, error = data.get_property_int("data")
        if error is not None:
            ten_env.log(LogLevel.ERROR, f"failed to get data: {error}")
            return

        substract, error = await ten_env.get_property_int("substract")
        if error is not None:
            ten_env.log(LogLevel.ERROR, f"failed to get substract: {error}")
            return

        result = number - substract
        data.set_property_int("data", result)
        await ten_env.send_data(data)
